### PR TITLE
Alerts CLI and general refactoring

### PR DESCRIPTION
**Pull request by camlorn**
I don't have the bandwidth to finish this before I'm off next week so we might want someone else to take it over.  This PR covers a sizable amount of refactoring, and a working alerts client.  To cut to the chase, the following two commands provision an alert (explanation of the pr follows):

```
./bin/morgue.js alerts target create --project cts \
--name test \
--workflow-name cts-alerts-test

./bin/morgue.js  alerts alert create --name test \
--query-period 1m \
--count fingerprint \
--trigger fingerprint,0,ge,5,10 \
--project cts \
--target-name test \
--enabled \
--min-notification-interval 1m \
--mute-until 0
```

In order to make this happen, and also so that this isn't a huge headache the next time we have to do one of these, we:

- Introduce eslint as a dev dependency, and use their default config to offer developers some optional local linting. I'm not intending to do more with this now, it's intentional that `npm run lint` doesn't specify paths because we can't run it on the whole codebase usefully, and we can figure out better settings later.
- Create a helper module for parsing, validating, and (in some instances) converting options.
- Create a helper base class for service clients, which offers functionality around talking to services in an authenticated way, and a pager which matches Rust
- Lift processing of folds out of `coronerList` in `bin/morgue.js` and into the new module for queries.
- Make our implicit query building steps around timestamps optional (but default on, for API compatibility).

This isn't entirely complete but this should be enough to unblock people. Specifically:

- I've only really tested the create paths.  Update and delete may be weak.
- We probably want some sort of better list, but in general don't have good formatting options for this data.
- There's no docs, and misusing the commands just prints `usage` as a placeholder for the time being.

If someone else doesn't pick this up, I'll address the documentation points and try to get some better testing in when I've got bandwidth again.  Overall, this should be reviewable as-is and it should be possible to work out what it's doing by just reading `lib/alerts/cli.js`, since the refactors underneath all of this get rid of most of the code duplication and leave only code related to alerts in that file.

There's some more targeted information in the individual commits.  I tried to split this up as we would with Phabricator.

### CODE REVIEW

File: lib/serviceClient.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+   */
+  request({ method, path, body = null, qs = {} }) {
+    let actualQs = {};
+    for (const [k, v] of Object.entries({ ...qs, ... this.defaultQs })) {
+      if (v === undefined || v == null) {
+        continue;
+      }
+      actualQs[k] = v;
+    }
+    return new Promise((resolve, reject) => {
```
Comment (konraddysput) [inline lib/serviceClient.js:54]: Hey- I didnt' spent too much time on this - but it looks like your request method will fail on on-premise environment, or overall with coronerd instances/service instances with self generated certificates.

File: lib/serviceClient.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ * we'll need to clean it up some.
+ *
+ * To use, subclass this class, pass the parameters up to the constructor,
+ * and then use this.request to make requests. The class will handle
+ * authenticating and manual url location.
+ */
+const request = require('request');
+const urlJoin = require('url-join');
+
+class ServiceClient {
```
Comment (konraddysput) [inline lib/serviceClient.js:16]: I really like the idea of standarizing things like that, however I'm against that without team discussion. Also `serviceClient` name suggests a name for `service client`  - something that all of our services can use to communicate with any type of service, but this class has a specific logic to your service.

**Review by konraddysput — CHANGES_REQUESTED** [review]

**Comment by camlorn** [comment]
I have pushed updates that should make it more clear that ServiceClient is supposed to be used as a base class, including a rename.  I've also handled `-k`, which is the CLI option that is supposed to be used in the self-signed cers scenario, by funneling it down from the Coroner client that is created to find the service.

I'm not going to rename BaseServiceClient to AlertsServiceClient since the alerts-specific logic is in AlertsClient in lib/alerts/client.js already, and the only piece of logic that it contains that is even arguably specific to alerts is the Rust pagination scheme, which applies to all Rust services, and is only coincidentally alerts specific because we haven't done a Rust service.  But I have expanded the docs to make it clear what it does, so that should be sufficient.  When I designed that interface, I was also doing it from the experience of just having done this for metrics importer, and intend to use this base class as-is in future, save perhaps with an additional pagination strategy or two added.

Also, needing to handle self-signed certs is an argument for the class and in fact an argument for generalizing further, though I don't intend to do that now: the tradeoff of making further changes here vs. manual testing time isn't worth it.

Monday, I'll find out about testing the self-signed certs paths.  I'm not sure how to do it without asking, but what I've done about that matches up with the documentation for request.

**Comment by konraddysput** [comment]
@camlorn, if you will install your service on my box, you should be able to reproduce the problem with self signed certificates. Lucky, that I have always problems with ssl certs there :p

**Comment by camlorn** [comment]
I've tested against a box with bad SSL certs, and it gets as far as an http 400 against the service for reasons I understand (in this case, needing to fully provision a user, etc).  I think this is good enough to prove that SSL is fine.  I've also pushed a couple fixes that this spotted along the way and think we're good to close this out, unless reviewers have further feedback.

Keep in mind that alerts is beta.  While this is a lot of code, it's currently a good time to do this sort of abstraction.

**Review by konraddysput — APPROVED** [review]

**Pull request merged**